api/islandora_xacml_api.drush.inc

```diff
@@ -135,3 +157,144 @@ EOQ;
     '@total' => $sandbox['total'],
   ));
 }
+
+/**
+ * Drush command callback for RELS-INT inconsistencies batch.
+ */
+function drush_islandora_xacml_api_relationships_missing_rels_int_batch() {
+  $batch = array(
+    'operations' => array(
+      array(
+        'islandora_xacml_api_relationships_missing_rels_int_update_batch_operation',
+        array(
+          drush_get_option('limit', 100),
+          drush_get_option('commit', FALSE),
+        ),
+      ),
+    ),
+  );
+
+  batch_set($batch);
+
+  drush_backend_batch_process();
+}
+
+/**
+ * Batch operation; identify and update a PID.
+ */
+function islandora_xacml_api_relationships_missing_rels_int_update_batch_operation($limit, $commit, &$context) {
+  if (!variable_get('islandora_xacml_api_save_relationships', TRUE)) {
+    drush_log(t('Relationships are disabled, nothing to do.'));
+  }
+  $sandbox =& $context['sandbox'];
+
+  // XXX: Can occasionally get be initialized as a different user...
+  drupal_static_reset('islandora_get_tuque_connection');
+  $tuque = islandora_get_tuque_connection();
+
+  $query = <<<EOQ
+PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+SELECT DISTINCT ?object ?created FROM <#ri>
+WHERE {
+  ?object <fedora-view:disseminates> ?ds ;
+          <fedora-view:disseminates> ?ds2 ;
+          <fedora-model:createdDate> ?created .
+  ?ds <fedora-view:disseminationType> <info:fedora/*/POLICY> .
+  {
+    ?ds2 <http://islandora.ca/ontology/relsint#isViewableByRole> "administrator"
+  }
+  UNION
+  {
+    ?ds2 <http://islandora.ca/ontology/relsint#isViewableByUser> "fedoraAdmin"
+  }
+  !filter
+}
+ORDER BY ?created
+
+EOQ;
+  if (!isset($sandbox['total'])) {
+    $sandbox['progress'] = 0;
+    $sandbox['total'] = $tuque->repository->ri->countQuery(format_string($query, array(
+      '!filter' => '',
+    )), 'sparql');
+    if ($sandbox['total'] == 0) {
+      $context['finished'] = 1;
+      $context['message'] = t('Nothing to verify.');
+      return;
+    }
+    $sandbox['result_stash'] = array();
+  }
+  if (empty($sandbox['result_stash'])) {
+    if (isset($sandbox['last_date'])) {
+      $filter = format_string('FILTER(?created > "!created"^^xsd:dateTime || (?created = "!created"^^xsd:dateTime && xsd:string(?object) > "info:fedora/!pid"^^xsd:string))', array(
```

Comment:
I don't completely understand the $filter logic here. Will try to dissect it and please correct me:
1. $query fetches all objects that have an attached policy and any indexed data stream with those two rels int properties (union) in case only one exists.

2.  at the end , limit is set to $LIMIT,the number of objects to process we pass to drush from drush_get_option('limit', 100),.  Lets say 100.
So it will get me every time 100.
3. The filter is set (in case it runs for the second time) to all objects created after $sandbox['last_date'] (accumulated on each batch run) OR   date is the same as $sandbox['last_date'] and the PID is larger/higher/next to the last PID processed? What makes you assume that PID will be actually higher if you are ordering by datecreate? and/or using UUID?
3. You run also an un-limited query (no $limit statement?) with the filter applied version at line 214 that gets us all (count) objects that match the query. 

So regarding the $filter and point 3(now i know what i don't get) not sure if checking the creationdate and the comparing a string via >pid is enough. Also, do all of our CMODELS get XACML inherited? 
Maybe dealing with modification dates.. in case the actual "fix" is being applied and it is not a dry run.

Reply:
Missing an additional order by `?pid` in the base query, should be `ORDER BY ?created ?pid` that is in the event that the ISO8601 stamp matches directly we'll fallback on the PID being ahead of the last one processed.

Not sure I fully understand your question(s) here but I'll lay out how this batch operates:

1. Batch is initiated and the total amount of objects that are to be processed is grabbed mainly for progress reporting / handling the end condition for the batch.
2. Initial chunk of objects are grabbed (without any filters) and processed, each storing their created date in the sandbox.
3. Initial chunk completes, next result set is grabbed using the created date of the last object processed for comparison purposes.

And so on.

Reply:
Hi, cool!. All good, the other 1-4 (except 3) were just my own dissecting to understand the hows and also to let other people reading get why my concern in 3. order by pid should solve the issue.